server/src/main/java/org/candlepin/controller/ActivationKeyController.java

```diff
@@ -0,0 +1,246 @@
+
+import com.google.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xnap.commons.i18n.I18n;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * ActivationKeyController
+ */
+public class ActivationKeyController {
+
+    private static final Logger log = LoggerFactory.getLogger(ActivationKeyController.class);
+    private static final Pattern AK_CHAR_FILTER = Pattern.compile("^[a-zA-Z0-9_-]+$");
+
+    private final ActivationKeyCurator activationKeyCurator;
+    private final OwnerProductCurator ownerProductCurator;
+    private final PoolManager poolManager;
+    private final I18n i18n;
+    private final ServiceLevelValidator serviceLevelValidator;
+    private final ActivationKeyRules activationKeyRules;
+
+    @Inject
+    public ActivationKeyController(ActivationKeyCurator activationKeyCurator, I18n i18n,
+        PoolManager poolManager, ServiceLevelValidator serviceLevelValidator,
+        ActivationKeyRules activationKeyRules, OwnerProductCurator ownerProductCurator) {
+
+        this.activationKeyCurator = Objects.requireNonNull(activationKeyCurator);
+        this.i18n = Objects.requireNonNull(i18n);
+        this.poolManager = Objects.requireNonNull(poolManager);
+        this.serviceLevelValidator = Objects.requireNonNull(serviceLevelValidator);
+        this.activationKeyRules = Objects.requireNonNull(activationKeyRules);
+        this.ownerProductCurator = Objects.requireNonNull(ownerProductCurator);
+    }
+
+    /**
+     * Fetches an activation key using the specified key ID. If a valid activation key could not be
+     * found, this method throws an exception.
+     *
+     * @param keyId
+     *  The ID of the activation key to fetch
+     *
+     * @throws IllegalArgumentException
+     *  if the given ID is null, empty
+     *
+     * @throws NotFoundException
+     *  if the given ID is not associated with a valid activation key
+     *
+     * @return
+     *  an ActivationKey with the specified ID
+     */
+    private ActivationKey fetchActivationKey(String keyId) {
+        if (keyId == null || keyId.isEmpty()) {
+            throw new IllegalArgumentException(i18n.tr("activation key ID is null or empty"));
+        }
+
+        ActivationKey key = this.activationKeyCurator.secureGet(keyId);
+
+        if (key == null) {
+            throw new NotFoundException(i18n.tr("ActivationKey with id {0} could not be found.", keyId));
+        }
+
+        return key;
+    }
+
+    public ActivationKey getActivationKey(String activationKeyId) {
+        return this.fetchActivationKey(activationKeyId);
+    }
+
+    public Set<ActivationKeyPool> getActivationKeyPools(String activationKeyId) {
+        return this.fetchActivationKey(activationKeyId)
+            .getPools();
+    }
+
+    public ActivationKey updateActivationKey(String activationKeyId, ActivationKeyDTO update) {
```

Comment:
It seems strange to have DTOs which are presentation layer information mixed in with the Controller which is acting on internal models. I would expect the validation of the ActivationKeyDTO & the conversion to an ActivationKey to be taking place in the Resource while the update action for the model is using ActivationKey takes place in the controller.